### ISSUE TITLE
color picker best practice

### ISSUE THREAD

**Issue by Ennoriel** [w3c/aria-practices#2742]
As the author of a color picker library, I'd like to make it as accessible as possible ([svelte awesome color picker](https://svelte-awesome-color-picker.vercel.app/#always-open-version)). However there is no such thing as a 2 dimension slider documented anywhere.

The thing is also that the most common color representation is the HSV one which is not very intuitive to explain in labels.

At the moment, the color picker is made of 2 distinct areas:
- a Hue, Saturation, Vibrance (brightness) and Alpha (transparency) pickers with the Saturation and Vibrance being a 2 dimensions picker.
- Inputs either hexadecimal, RGB or HSV.

I have troubles with the 2 dimensions picker to compute the `aria-min`, `aria-max`, `aria-valuenow` and `aria-valuetext` properties. The role and keyboard events are also not standard...

I am thinking of dividing the picker in 2 distinct keyboard selectable sliders for each property but would that be better? What would you suggest otherwise? Would that fit in the design patterns?

I have a [work in progress version of the picker](https://svelte-awesome-color-picker-git-feature-contrast-level-ennoriel.vercel.app/open) where I display the 3:0, 4.5:0 and 7:0 contrast levels within the picker and really don't know how to make it accessible the right way.

**Comment by JAWS-test** [comment]
A two-dimensional slider is currently not defined. I see four possibilities:
1. split on two one-dimensional sliders.
2. two-dimensional slider with operation hint (aria-describedby) that arrow key operation is possible in 2 dimensions and use of aria-valuetext to output the values
3. addition to ARIA specification of slider, so that two-dimensional slider is allowed
4. additional alternative input possibility to the slider: input of values via [spin button](https://www.w3.org/WAI/ARIA/apg/patterns/spinbutton/examples/datepicker-spinbuttons/)

Please note:
* operation hints should not be in the aria-label (you have the hint for arrow key operation there).
* the normal sliders do not need a hint for arrow key operation (you currently have it there)
* the colorpicker should not be opened automatically when the buttons are focused, but only after activating the buttons - a keyboard user might want to navigate over the button and not be "trapped" in the colorpicker
* the colorpicker should be able to be closed with ESC
* for blind users the exceeding of a limit value concerning AA or AAA WCAG contrast requirements can be transmitted e.g. via Live Region

**Comment by JAWS-test** [comment]
I don't know if the different implementation was done on purpose. I think it doesn't matter where the role=slider is located. It only matters that role=slider is at the same element that gets the keyboard focus. 

* Advantage of role=slider on the whole element and not only on the thumb is that there is also a correct output on mouseover (e.g. when using a screen magnifier or other speech output used in conjunction with mouse operation). 
* The advantage of role=slider on the thumb could be that the thumb could be displayed automatically in the visible area more easily at large zoom levels (see WCAG 1.4.4 and 1.4.10) when using the keyboard (unfortunately, this is not currently the case for both examples from the W3C).

**Comment by JAWS-test** [comment]
> I am not sure to know what it means. Could you please explain it?

When the slider is operated, the system automatically displays whether the WCAG contrast requirements are met. This is not perceptible with a screen reader. A live region can therefore be used: https://w3c.github.io/aria/#aria-live

**Comment by JAWS-test** [comment]
At https://svelte-awesome-slider.vercel.app/ the contrasts of the slider bar are not sufficient. e.g. first example: left light blue to white 1.3:1, right 2.1:1.

When operating the slider with page up/down or Pos1/End, the page is scrolled and the slider is no longer visible

**Comment by JAWS-test** [comment]
At https://svelte-awesome-color-picker-git-v300-ennoriel.vercel.app/ the contrasts are also insufficient (e.g. white text on orange for insufficient contrast has only 2:1; form fields do not have sufficient contrast)

If the pop-up is opened with the colour picker, the focus is not set in the pop-up. In addition, the keyboard focus does not remain in the pop-up.

The buttons to change the colour model are well labelled for blind people, but visual labels do not match, which is a violation of WCAG 2.5.3

**Comment by Ennoriel** [comment]
Thanks for all the feedbacks!

I took them into account for both libraries:

- https://svelte-awesome-slider.vercel.app/
- https://svelte-awesome-color-picker-git-v300-ennoriel.vercel.app/

Don't hesitate to let me know if I implemented them correctly or if you find something else that could be improved!

However, I did not understand how to use aria-live:

> When the slider is operated, the system automatically displays whether the WCAG contrast requirements are met. This is not perceptible with a screen reader. A live region can therefore be used: https://w3c.github.io/aria/#aria-live

Would you have an example of how to use it?

**Comment by JAWS-test** [comment]
> Don't hesitate to let me know if I implemented them correctly or if you find something else that could be improved!

I won't look at the examples again, firstly I don't have time for that, secondly this is not the place for individual advice and thirdly you might get too one-sided a picture if only I comment - actually others should take part in the discussion

**Comment by JAWS-test** [comment]
> Would you have an example of how to use it?

https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/ARIA_Live_Regions

Important for live regions: The (usually empty) live region must already be in the DOM (e.g. `<div aria-live=polite></div>`). A short text is then inserted into the live region using JavaScript ("Contrast too low", so that the DOM then looks like this: <div aria-live=polite>Contrast too low</div>).

**Comment by Ennoriel** [comment]
Thanks for the documentation and the many advice you gave me.

I am not an expert with this at all. Working with live regions is tough. I'm not very confident about how I should implement it. With one of the speaker I've installed, the live region is sometimes announced before the value change, and sometimes after and I didn't succed to find a way to announce not too much change and not too little by still being clear about what is meant by one of the contrast too low / good enough.

I'm still thinking...

**Comment by JAWS-test** [comment]
Yes, live regions do not work very reliably and consistently. Then simply leave out the live region and offer a way of using the screen reader to see in text form which WCAG thresholds are not met
